src/execution/sql/table_vector_iterator.cpp

```diff
@@ -1,63 +1,206 @@
+  // TODO(WAN): code duplication. Though it does make it pretty clear.
+
+  // No-op if already initialized
+  if (IsInitialized()) {
+    return true;
+  }
+
+  // Set up the table and the iterator.
+  table_ = exec_ctx_->GetAccessor()->GetTable(table_oid_);
+  TERRIER_ASSERT(table_ != nullptr, "Table must exist!!");
+  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->GetBlockedSlotIterator(block_start, block_end));
+  const auto &table_col_map = table_->GetColumnMap();
+
+  // Configure the vector projection, create the column iterators.
+  std::vector<storage::col_id_t> col_ids;
+  std::vector<TypeId> col_types(col_oids_.size());
+  column_iterators_.reserve(col_oids_.size());
+  for (uint64_t idx = 0; idx < col_oids_.size(); idx++) {
+    auto col_oid = col_oids_[idx];
+    auto col_type = GetTypeId(table_col_map.at(col_oid).col_type_);
+    auto storage_col_id = table_col_map.at(col_oid).col_id_;
+
+    col_ids.emplace_back(storage_col_id);
+    col_types[idx] = col_type;
+    column_iterators_.emplace_back(GetTypeIdSize(col_type));
+  }
+
+  // Create an owning vector.
+  vector_projection_.SetStorageColIds(col_ids);
+  vector_projection_.Initialize(col_types);
+  vector_projection_.Reset(common::Constants::K_DEFAULT_VECTOR_SIZE);
+
+  // All good.
+  initialized_ = true;
   return true;
 }
 
+void TableVectorIterator::RefreshVectorProjection() {
+  // Reset our projection and refresh all columns with new data from the column iterators.
+
+  const uint32_t tuple_count = column_iterators_[0].GetTupleCount();
+
+  TERRIER_ASSERT(std::all_of(column_iterators_.begin(), column_iterators_.end(),
+                             [&](const auto &iter) { return tuple_count == iter.GetTupleCount(); }),
+                 "Not all iterators have the same size?");
+
+  vector_projection_.Reset(tuple_count);
+  for (uint64_t col_idx = 0; col_idx < column_iterators_.size(); col_idx++) {
+    Vector *column_vector = vector_projection_.GetColumn(col_idx);
+    column_vector->Reference(column_iterators_[col_idx].GetColumnData(),
+                             column_iterators_[col_idx].GetColumnNullBitmap(),
+                             column_iterators_[col_idx].GetTupleCount());
+  }
+  vector_projection_.CheckIntegrity();
+
+  // Insert our vector projection instance into the vector projection iterator.
+  vector_projection_iterator_.SetVectorProjection(&vector_projection_);
+}
+
 bool TableVectorIterator::Advance() {
-  if (!initialized_) return false;
-  // First check if the iterator ended.
-  if (*iter_ == table_->end()) {
+  // Cannot advance if not initialized.
+  if (!IsInitialized()) {
     return false;
   }
-  // Scan the table to set the projected column.
-  table_->Scan(exec_ctx_->GetTxn(), iter_.get(), projected_columns_);
-  pci_.SetProjectedColumn(projected_columns_);
+
+  // If the iterator is out of data, then we are done.
+  if (*iter_ == table_->end() || (**iter_).GetBlock() == nullptr) {
```

Comment:
Is it possible to have a valid iterator without a block?

Reply:
Need to check. I don't think so?